File: src/main/java/cc/carm/plugin/intellij/quarkdown/actions/BaseToggleAction.java
package cc.carm.plugin.intellij.quarkdown.actions;

import com.intellij.openapi.actionSystem.AnAction;
import com.intellij.openapi.actionSystem.AnActionEvent;
import com.intellij.openapi.actionSystem.CommonDataKeys;
import com.intellij.openapi.command.WriteCommandAction;
import com.intellij.openapi.editor.Caret;
import com.intellij.openapi.editor.Document;
import com.intellij.openapi.editor.Editor;
import com.intellij.openapi.project.DumbAware;
import org.jetbrains.annotations.NotNull;

public abstract class BaseToggleAction extends AnAction implements DumbAware {

    @Override
    public void update(@NotNull AnActionEvent e) {
        Editor editor = e.getData(CommonDataKeys.EDITOR);
        e.getPresentation().setEnabled(editor != null && !editor.isViewer());
    }

    @Override
    public void actionPerformed(@NotNull AnActionEvent e) {
        Editor editor = e.getData(CommonDataKeys.EDITOR);
        if (editor == null) return;
        Document document = editor.getDocument();

        WriteCommandAction.runWriteCommandAction(e.getProject(), () -> {
            for (Caret caret : editor.getCaretModel().getAllCarets()) {
                if (caret.hasSelection()) {
                    toggleSelection(document, caret);
                } else {
                    insertWrappers(document, caret);
                }
            }
        });
    }

    protected abstract @NotNull String getWrapper();

    private void toggleSelection(Document document, Caret caret) {
        int start = caret.getSelectionStart();
        int end = caret.getSelectionEnd();
        String selected = document.getText().substring(start, end);
        String wrapper = getWrapper();
        int wrapperLen = wrapper.length();

        if (selected.startsWith(wrapper) && selected.endsWith(wrapper) && selected.length() >= wrapperLen * 2) {
            String inner = selected.substring(wrapperLen, selected.length() - wrapperLen);
            document.replaceString(start, end, inner);
            caret.removeSelection();
            caret.moveToOffset(start + inner.length());
        } else {
            String wrapped = wrapper + selected + wrapper;
            document.replaceString(start, end, wrapped);
            caret.setSelection(start, start + wrapped.length());
        }
    }

    private void insertWrappers(Document document, Caret caret) {
        String wrapper = getWrapper();
        document.insertString(caret.getOffset(), wrapper + wrapper);
        caret.moveCaretRelatively(wrapper.length(), 0, false, false);
    }
}

File: src/main/kotlin/cc/carm/plugin/intellij/quarkdown/actions/BaseToggleAction.kt
package cc.carm.plugin.intellij.quarkdown.actions

import com.intellij.openapi.actionSystem.ActionUpdateThread
import com.intellij.openapi.actionSystem.AnAction
import com.intellij.openapi.actionSystem.AnActionEvent
import com.intellij.openapi.actionSystem.CommonDataKeys
import com.intellij.openapi.command.WriteCommandAction
import com.intellij.openapi.editor.Document
import com.intellij.openapi.project.DumbAware

abstract class BaseToggleAction : AnAction(), DumbAware {

    override fun getActionUpdateThread() = ActionUpdateThread.BGT

    abstract fun getWrapper(): String

    override fun update(e: AnActionEvent) {
        val editor = e.getData(CommonDataKeys.EDITOR)
        e.presentation.isEnabled = editor != null && !editor.isViewer
    }

    override fun actionPerformed(e: AnActionEvent) {
        val editor = e.getData(CommonDataKeys.EDITOR) ?: return
        val document = editor.document

        WriteCommandAction.runWriteCommandAction(e.project) {
            for (caret in editor.caretModel.allCarets) {
                if (caret.hasSelection()) {
                    toggleSelection(document, caret)
                } else {
                    insertWrappers(document, caret)
                }
            }
        }
    }

    private fun toggleSelection(document: Document, caret: com.intellij.openapi.editor.Caret) {
        val start = caret.selectionStart
        val end = caret.selectionEnd
        val selected = document.text.substring(start, end)
        val wrapper = getWrapper()
        val wrapperLen = wrapper.length

        if (selected.startsWith(wrapper) && selected.endsWith(wrapper) && selected.length >= wrapperLen * 2) {
            val inner = selected.substring(wrapperLen, selected.length - wrapperLen)
            document.replaceString(start, end, inner)
            caret.removeSelection()
            caret.moveToOffset(start + inner.length)
        } else {
            val wrapped = wrapper + selected + wrapper
            document.replaceString(start, end, wrapped)
            caret.setSelection(start, start + wrapped.length)
        }
    }

    private fun insertWrappers(document: Document, caret: com.intellij.openapi.editor.Caret) {
        val wrapper = getWrapper()
        document.insertString(caret.offset, wrapper + wrapper)
        caret.moveCaretRelatively(wrapper.length, 0, false, false)
    }
}
